From 784c3555a5330ef8329c0d0cf4757d57933d3d5b Mon Sep 17 00:00:00 2001
From: Panu Matilainen <pmatilai@redhat.com>
Date: Wed, 3 Dec 2014 07:44:00 -0500
Subject: [net] gre: Fix use-after-free panic in ipgre_rcv()

Message-id: <493e2912a2796d5be60e06f80b8f22f52f1bb2c3.1417529969.git.pmatilai@redhat.com>
Patchwork-id: 101102
O-Subject: [RHEL7 PATCH net] gre: Fix use-after-free panic in ipgre_rcv()
Bugzilla: 1117543
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1117543
Upstream: fbd02dd405d0724a0f25897ed4a6813297c9b96f
Tested: Locally in VM, reproducer in bugzilla
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8318594

Check for looped-back packet before its released, simple backport
from upstream. The RHEL-7 kernel never had the buggy intermediate
fix from commit 10ddceb22ba so the skb_dst_drop() doesn't need
to be added back.

Backport of commit fbd02dd405d0724a0f25897ed4a6813297c9b96f:
commit fbd02dd405d0724a0f25897ed4a6813297c9b96f
Author: Pravin B Shelar <pshelar@nicira.com>
Date:   Sun Mar 23 22:06:36 2014 -0700

    ip_tunnel: Fix dst ref-count.

    Commit 10ddceb22ba (ip_tunnel:multicast process cause panic due
    to skb->_skb_refdst NULL pointer) removed dst-drop call from
    ip-tunnel-recv.

    Following commit reintroduce dst-drop and fix the original bug by
    checking loopback packet before releasing dst.
    Original bug: https://bugzilla.kernel.org/show_bug.cgi?id=70681

    CC: Xin Long <lucien.xin@gmail.com>
    Signed-off-by: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Panu Matilainen <pmatilai@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/gre_demux.c b/net/ipv4/gre_demux.c
index 9c8333f..7e0756d 100644
--- a/net/ipv4/gre_demux.c
+++ b/net/ipv4/gre_demux.c
@@ -166,6 +166,14 @@ static int gre_cisco_rcv(struct sk_buff *skb)
  int i;
  bool csum_err = false;
 
+#ifdef CONFIG_NET_IPGRE_BROADCAST
+ if (ipv4_is_multicast(ip_hdr(skb)->daddr)) {
+  /* Looped back packet, drop it! */
+  if (rt_is_output_route(skb_rtable(skb)))
+   goto drop;
+ }
+#endif
+
  if (parse_gre_header(skb, &tpi, &csum_err) < 0)
   goto drop;
 
diff --git a/net/ipv4/ip_tunnel.c b/net/ipv4/ip_tunnel.c
index b278895..b1eeb95 100644
--- a/net/ipv4/ip_tunnel.c
+++ b/net/ipv4/ip_tunnel.c
@@ -415,9 +415,6 @@ int ip_tunnel_rcv(struct ip_tunnel *tunnel, struct sk_buff *skb,
 
 #ifdef CONFIG_NET_IPGRE_BROADCAST
  if (ipv4_is_multicast(iph->daddr)) {
-  /* Looped back packet, drop it! */
-  if (rt_is_output_route(skb_rtable(skb)))
-   goto drop;
   tunnel->dev->stats.multicast++;
   skb->pkt_type = PACKET_BROADCAST;
  }
-- 
1.7.1